source/fem-dem/gls_sharp_navier_stokes.cc

```diff
@@ -126,65 +131,107 @@ GLSSharpNavierStokesSolver<dim>::generate_cut_cells_map()
-              // the cell is cut.
-              if (nb_dof_inside != 0)
+              else
                 {
-                  // If all the DOFs are inside the boundary this cell is inside
-                  // the particle. Otherwise the particle is cut.
-                  if (nb_dof_inside == dofs_per_cell_local_v_x)
+                  unsigned int nb_dof_inside = 0;
+                  for (unsigned int j = 0; j < local_dof_indices.size(); ++j)
                     {
-                      // We only register the particle the lowest id as the
-                      // particle in which this cell is embedded if the cell is
-                      // embedded in multiple particles.
-                      if (number_of_particles_cutting_this_cell == 0)
+                      // Count the number of DOFs that are inside
+                      // of the particles. If all the DOfs are on one side
+                      // the cell is not cut by the boundary.
+                      if (0 == this->fe->system_to_component_index(j).first)
                         {
-                          cell_is_cut                                = false;
-                          particle_id_which_cuts_this_cell           = 0;
-                          cell_is_inside                             = true;
-                          particle_id_in_which_this_cell_is_embedded = p;
+                          // Check if we already have checked this point in a
+                          // previous cell. If we did we use the results that we
+                          // have previously obtained.
+                          auto iterator =
+                            inside_outside_support_point_vector[p].find(
+                              local_dof_indices[j]);
+                          if (iterator ==
+                              inside_outside_support_point_vector[p].end())
+                            {
+                              if (particles[p].get_levelset(
+                                    support_points[local_dof_indices[j]],
+                                    cell) <= 0)
+                                {
+                                  ++nb_dof_inside;
+                                  inside_outside_support_point_vector
+                                    [p][local_dof_indices[j]] = true;
+                                }
+                              else
+                                inside_outside_support_point_vector
+                                  [p][local_dof_indices[j]] = false;
+                            }
+                          else
+                            {
+                              if (inside_outside_support_point_vector
+                                    [p][local_dof_indices[j]] == true)
+                                {
+                                  ++nb_dof_inside;
+                                }
+                            }
```

Comment:
Here the code does the reverse of what you say. Because if it's equal to end() it means you did not find it. I would switch the code the other way around (treat the if  !=end()  before) Or at least adapt the comments.